extendr-api/src/wrapper/matrix.rs

```diff
@@ -351,6 +360,41 @@ impl<T, D> Deref for RArray<T, D> {
     }
 }
 
+trait CollectRMatrix<T> {
+    fn collect_rmatrix(self, nrow: usize) -> Result<RMatrix<T>>;
+}
+
+impl<'a, T, U> CollectRMatrix<U> for T
+where
+    T: Iterator<Item = U>,
+    U: ToVectorValue + 'a,
+    Robj: AsTypedSlice<'a, U>,
+{
+    /// Collects an iterable into an RMatrix.
+    /// The iterable must yield items column by column (aka Fortan order)
+    ///
+    /// # Arguments
+    ///
+    /// * `nrow` - the number of rows the matrix will have
+    fn collect_rmatrix(self, nrow: usize) -> Result<RMatrix<U>> {
+        let vector = self.collect_robj();
+        let ncol = vector.len() / nrow;
+        if ncol * nrow != vector.len() {
+            return Err(Error::Other(format!(
+                "The vector length ({}) is not divisible by nrow ({})",
+                vector.len(),
+                nrow
+            )));
+        }
+        let dim = [nrow, ncol];
+        let mut robj = vector
+            .set_attrib(wrapper::symbol::dim_symbol(), dim)
+            .unwrap();
+        let data = robj.as_typed_slice_mut().unwrap().as_mut_ptr();
```

Comment:
Is this never, ever going to fail or is it possible for this to fail?

Reply:
Good point, I'll `?` this.

Reply:
I looked at the source code for as_typed_slice_mut. It is infalliable. I don't know why it is. Just leave the unwrap